File: 121-Best-Time-to-Buy-and-Sell-Stock/solution.ts
function maxProfit(prices: number[]): number {
  let max = 0;
  let localMin = Number.MAX_SAFE_INTEGER;
  let localMinIndex = -1;
  for (let i = 0; i < prices.length; i++) {
    if (localMin> prices[i]){
        localMin = prices[i]
        localMinIndex = i
    }
    
    for (let j = localMinIndex; j < prices.length; j++) {
        const newPrice = prices[i];
        const oldPrice =  prices[j];
        
        const profit = newPrice-oldPrice
        max = Math.max(max,profit)
        
    }
  }
  return max;
}

File: 121-Best-Time-to-Buy-and-Sell-Stock/solution.java
class Solution {
    public int maxProfit(int[] prices) {
        
        int max = 0;
        int lIndex = 0;
        int rIndex = 1;
        for (int i = 0; i < prices.length && rIndex< prices.length; i++) {
            if (prices[lIndex]> prices[i]){
                lIndex = i;
            }
            int profit = prices[rIndex]-prices[lIndex];
            max = Math.max(max,profit);
            rIndex++;
        }
        return max;
    }
}